fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/directory/FDBDirectory.java

```diff
@@ -157,29 +159,56 @@ public FDBDirectory(@Nonnull Subspace subspace, @Nonnull FDBRecordContext contex
         this.encryptionEnabled = Objects.requireNonNullElse(context.getPropertyStorage().getPropertyValue(LuceneRecordContextProperties.LUCENE_INDEX_ENCRYPTION_ENABLED), false);
         this.fileReferenceMapSupplier = Suppliers.memoize(this::loadFileReferenceCacheForMemoization);
     }
 
+    private long deserializeFileSequenceCounter(@Nullable byte[] value) {
+        return value == null ? 0L : Tuple.fromBytes(value).getLong(0);
+    }
+
+    @Nonnull
+    private byte[] serializeFileSequenceCounter(long value) {
+        return Tuple.from(value).pack();
+    }
+
+    @Nonnull
+    private CompletableFuture<Void> loadFileSequenceCounter() {
+        long originalValue = fileSequenceCounter.get();
+        if (originalValue >= 0) {
+            // Already loaded
+            return AsyncUtil.DONE;
+        }
+        return context.ensureActive().get(sequenceSubspaceKey).thenAccept(serializedValue -> {
+            // Replace the counter value in the counter unless something else has already
+            // updated it, in which case we want to keep the cached value
+            long loadedValue = deserializeFileSequenceCounter(serializedValue);
+            fileSequenceCounter.compareAndSet(originalValue, loadedValue);
+        });
+    }
+
     /**
      * Sets increment if not set yet and waits till completed to return
      * Returns and increments the increment if its already set.
      * @return current increment value
      */
-    public synchronized long getIncrement() {
+    public long getIncrement() {
         context.increment(LuceneEvents.Counts.LUCENE_GET_INCREMENT_CALLS);
-        CompletableFuture<Long> future = context.ensureActive().get(sequenceSubspaceKey).thenApply(value -> {
-            if (value == null) {
-                context.ensureActive().set(sequenceSubspaceKey, Tuple.from(1L).pack());
-                return 1L;
-            } else {
-                long sequence = Tuple.fromBytes(value).getLong(0) + 1;
-                context.ensureActive().set(sequenceSubspaceKey, Tuple.from(sequence).pack());
-                return sequence;
-            }
-        });
-        return Objects.requireNonNull(context.asyncToSync(LuceneEvents.Waits.WAIT_LUCENE_GET_INCREMENT, future));
+
+        // Make sure the file counter is loaded from the database into the fileSequenceCounter, then
+        // use the cached value
+        context.asyncToSync(LuceneEvents.Waits.WAIT_LUCENE_GET_INCREMENT, loadFileSequenceCounter());
+        long incrementedValue = fileSequenceCounter.incrementAndGet();
+
+        // Use BYTE_MAX here so that if there are concurrent calls (which will get different increments because they
+        // are serialized around the fileSequenceCounter AtomicLong), the largest one always gets
+        // committed into the database
+        byte[] serializedValue = serializeFileSequenceCounter(incrementedValue);
+        context.ensureActive().mutate(MutationType.BYTE_MAX, sequenceSubspaceKey, serializedValue);
```

Comment:
Could this mutation fail but the counter incrementation above succeed?

Reply:
Not in a way that I think would cause a problem. The `mutate` method can fail (if, for example, the transaction has already been committed or closed), but only in ways that would, I think, not result in any data being committed based on the uncommitted data

Reply:
But does this cause an incorrect value to be updated to the counter? The failed ```mutate``` won't be handled with undoing the counter update above right?

Reply:
I think the answer is that, yes, that could theoretically result in a value in the counter that was out of sync with the value in the underlying transaction, but my point is that that doesn't matter. This value is only used internally in the `FDBDirectory`, not exposed to the user, so the only problematic case would be if, somehow, the value in the database's counter wasn't updated, but a value was somehow used as the file ID for some file. But if there is file data committed which uses this counter, then this `mutate` line must also have succeeded. (Because the `mutate` can't fail in a way that still allows subsequent writes to succeed.)